distributed/core.py

```diff
@@ -109,6 +110,105 @@ def _expects_comm(func: Callable) -> bool:
+
+    def call_soon(
+        self, afunc: CoroutineFunctionType, *args, **kwargs
+    ) -> asyncio.Task | None:
+        """Schedule a coroutine function to be executed as an `asyncio.Task`.
+
+        The coroutine function `afunc` is scheduled with `args` arguments and `kwargs` keyword arguments
+        as an `asyncio.Task`.
+
+        Parameters
+        ----------
+        afunc
+            Coroutine function to schedule.
+        *args
+            Arguments to be passed to `afunc`.
+        **kwargs
+            Keyword arguments to be passed to `afunc`
+
+        Returns
+        -------
+        asyncio.Task | None
+            The scheduled Task object, or None if the group is closed.
+        """
+        if self.closed:
+            return None
+
+        task = asyncio.create_task(afunc(*args, **kwargs))
+        self._ongoing_tasks.add(task)
+        task.add_done_callback(self._ongoing_tasks.remove)
+        return task
+
+    def call_later(
+        self, delay: int, afunc: CoroutineFunctionType, *args, **kwargs
+    ) -> asyncio.Task | None:
+        """Schedule a coroutine function to be executed after `delay` seconds as an `asyncio.Task`.
+
+        The coroutine function `afunc` is scheduled with `args` arguments and `kwargs` keyword arguments
+        as an `asyncio.Task` that is executed after `delay` seconds.
+
+        Parameters
+        ----------
+        delay
+            Delay in seconds.
+        afunc
+            Coroutine function to schedule.
+        *args
+            Arguments to be passed to `afunc`.
+        **kwargs
+            Keyword arguments to be passed to `afunc`
+
+        Returns
+        -------
+        asyncio.Task | None
+            The scheduled Task object, or None if the group is closed.
+        """
+        return self.call_soon(delayed(afunc, delay), *args, **kwargs)
+
+    @property
+    def _cancellable_tasks(self):
+        return (t for t in self._ongoing_tasks if t is not asyncio.current_task())
+
+    def close(self) -> None:
+        """Closes the task group so that no new tasks can be scheduled.
+
+        Existing tasks continue to run.
+        """
+        self.closed = True
+
+    async def stop(self, timeout=1) -> None:
+        """Close the group and stop all currently running tasks.
+
+        Closes the task group and waits `timeout` seconds for all tasks to gracefully finish.
+        After the timeout, all remaining tasks are cancelled.
+        """
+        self.close()
+
+        # Wrap gather in task to avoid Python3.8 issue,
+        # see https://github.com/dask/distributed/pull/6478#discussion_r885757056
+        gather_task = asyncio.create_task(
+            asyncio.gather(*self._cancellable_tasks, return_exceptions=True)
+        )
+        try:
+            await asyncio.wait_for(
+                gather_task,
+                timeout,
+            )
+        except asyncio.TimeoutError:
+            pass
```

Comment:
We have at least two places where we'd need to use the same logic if we were to drop `TaskGroup`, namely for dealing with `_ongoing_comm_handlers` and `_ongoing_background_tasks`. That's why I think it might be worthwhile to keep the `TaskGroup` class around. IMO it is also pretty well encapsulated, so it makes the code more readable than a flattened version.